### PR TITLE
Migrate dartdoc_options.dart

### PR DESCRIPTION
**Pull request by jcollins-g**
This one was tricksy because it involves argument parsing and yaml parsing, two areas that are null minefields.

Significant nullability choices:

- `.parent` is mostly non-nullable except for a specially defined root node that will throw if you try to access its parent.  Makes things easier to read, but the way that trees of options are constructed doesn't quite sit right with me now -- those `late` declarations hang around too long before being written to.  Added a TODO around that as I don't have a good alternative just yet.
- `valueAt` is nullable.  It's just too hard to try to restrict that given how `package:args` works and yaml giving us nulls.  If we happen to know some values aren't nullable, we can wrap them in the future.
- `operator []` is non-nullable.  I think this is the right choice as child nodes in the option tree are defined by the programmer, and we should never be trying to look up an undefined option.  Changing this found some bugs, too.

Most everything else follows naturally from these choices and from #2745.

### CODE REVIEW

**Comment by jcollins-g** [comment]
Also, the tests are not migrated; after running into trouble and reading [this](https://dart.dev/null-safety/unsound-null-safety#testing-or-running-mixed-version-programs), I've decided trying to work around the problem of entry points triggering strong nullability checks by default in tests is too much to be worth it.  So we'll just settle for a Test Skitching Development style.

File: lib/src/dartdoc_options.dart [hunk 1]
```diff
@@ -92,12 +90,10 @@ class CategoryConfiguration {
     var newCategoryDefinitions = <String, CategoryDefinition>{};
     for (var entry in yamlMap.entries) {
       var name = entry.key.toString();
-      String displayName;
-      String documentationMarkdown;
       var categoryMap = entry.value;
       if (categoryMap is Map) {
-        displayName = categoryMap['displayName']?.toString();
-        documentationMarkdown = categoryMap['markdown']?.toString();
+        var displayName = categoryMap['displayName']?.toString();
```
Comment (srawlins) [inline lib/src/dartdoc_options.dart:95]: haha easy win

File: lib/src/dartdoc_options.dart [hunk 2]
```diff
@@ -122,9 +118,7 @@ class ToolConfiguration {
 
   final ResourceProvider resourceProvider;
 
-  ToolRunner _runner;
-
-  ToolRunner get runner => _runner ??= ToolRunner(this);
+  late ToolRunner runner = ToolRunner(this);
```
Comment (srawlins) [inline lib/src/dartdoc_options.dart:121]: slick

File: lib/src/dartdoc_options.dart [hunk 3]
```diff
@@ -298,10 +292,9 @@ class _OptionValueWithContext<T> {
   ///
   /// [path] is the path where this value came from (not required to be
   /// canonical).
-  _OptionValueWithContext(this.value, String path, {this.definingFile}) {
-    canonicalDirectoryPath = p.canonicalize(path);
-    pathContext = p.Context(current: canonicalDirectoryPath);
-  }
+  _OptionValueWithContext(this.value, String path, {this.definingFile})
+      : canonicalDirectoryPath = p.canonicalize(path),
```
Comment (srawlins) [inline lib/src/dartdoc_options.dart:296]: +1 nice refactor

File: lib/src/dartdoc_options.dart [hunk 4]
```diff
@@ -432,14 +407,15 @@ abstract class DartdocOption<T> {
...
+    var resolvedPaths = <String>[];
     var value = valueWithContext.value;
     if (value is String) {
       resolvedPaths = [valueWithContext.resolvedValue as String];
     } else if (value is List<String>) {
       resolvedPaths = valueWithContext.resolvedValue as List<String>;
     } else if (value is Map<String, String>) {
-      resolvedPaths = (valueWithContext.resolvedValue as Map).values.toList();
+      resolvedPaths = (valueWithContext.resolvedValue as Map).values.toList()
+          as List<String>;
```
Comment (srawlins) [inline lib/src/dartdoc_options.dart:418]: I wonder if a spread formats nicer? Something like

    resolvedPaths = [
        ...(valueWithContext.resolvedValue as Map).values,
    ]

Not sure if there is a type error there, or if it actually would format nicer...

File: lib/src/dartdoc_options.dart [hunk 5]
```diff
@@ -705,115 +657,133 @@ abstract class DartdocSyntheticOption<T> implements DartdocOption<T> {
...
-class DartdocOptionSet extends DartdocOption<void> {
-  DartdocOptionSet(String name, ResourceProvider resourceProvider)
-      : super(
-            name, null, null, OptionKind.other, false, null, resourceProvider);
+/// This is a [DartdocOptionSet] used as a root node.
+class DartdocOptionRoot extends DartdocOptionSet {
+  DartdocOptionRoot(String name, ResourceProvider resourceProvider)
+      : super(name, resourceProvider);
+
+  late final ArgParser __argParser =
```
Comment (srawlins) [inline lib/src/dartdoc_options.dart:665]: Is the double underscore a pattern from before the null safety migration, or new?

**Review by srawlins — APPROVED** [review]
Love it

**Comment by srawlins** [comment]
~~Quick question~~ Big question: have you benchmarked along the null safety migration?

I'd be curious to know if the time-to-document Flutter has changed (I doubt it would), and more curious to know if the lazy initialization changes perhaps lead to different max memory usage. I think I doubt this too because theoretically the `late final` laziness is exactly the same as the explicit memoization laziness, but in practice...

File: lib/src/dartdoc_options.dart (hunk 5 above)
Comment (jcollins-g) [reply lib/src/dartdoc_options.dart:665]: It's from before.  I wrote this when I was still trying to write Python in Dart.  I can change.

**Comment by jcollins-g** [comment]
I haven't benchmarked it yet; was planning to though when the NNBD migration is a little closer

File: lib/src/dartdoc_options.dart (hunk 1 above)
Comment (jcollins-g) [reply lib/src/dartdoc_options.dart:95]: Yeah, my early Dart code was a little cringey, I'm happy to give it a touch up as part of this migration :-)

File: lib/src/dartdoc_options.dart (hunk 4 above)
Comment (jcollins-g) [reply lib/src/dartdoc_options.dart:418]: Ah yes, spread format is win here, good call

**Pull request merged**